File: 151051/Maliuk/ipr/src/main/java/ru/bsuir/services/PostService.java
package ru.bsuir.services;

import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Cacheable;
import lombok.AllArgsConstructor;
import org.springframework.cache.annotation.EnableCaching;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;
import ru.bsuir.dto.request.PostRequestTo;
import ru.bsuir.dto.response.PostResponseTo;
import ru.bsuir.entity.Post;
import ru.bsuir.entity.Story;
import ru.bsuir.irepositories.IPostRepository;
import ru.bsuir.irepositories.IStoryRepository;
import ru.bsuir.mapper.PostMapper;

import java.util.List;
import java.util.Optional;

@Service
@AllArgsConstructor
@EnableCaching
public class PostService {

    private final IPostRepository postRepository;

    private final PostMapper postMapper;

    private final IStoryRepository postRepository;

    @CacheEvict(value = {"posts", "postsList"}, allEntries = true)
    public PostResponseTo createPost(PostRequestTo postRequest) {

        Story story = postRepository.findById(postRequest.getTweetId())
                .orElseThrow(() -> new DataIntegrityViolationException("Story not found"));

        Post post = postMapper.toEntity(postRequest);
        post.setStory(story);
        post = postRepository.save(post);
        return postMapper.toDTO(post);
    }

    @Cacheable(value = "posts", key = "#id")
    public PostResponseTo getPostById(Long id) {
        Optional<Post> post = postRepository.findById(id);
        return post.map(postMapper::toDTO)
                .orElseThrow(() -> new RuntimeException("Post not found"));
    }

    @Cacheable(value = "postsList")
    public List<PostResponseTo> getAllPost(){
        return postRepository.findAll().stream()
                .map(postMapper::toDTO)
                .toList();
    }

    @CacheEvict(value = {"posts", "cpostsList"}, key = "#id", allEntries = true)
    public PostResponseTo updatePost(Long id, PostRequestTo postRequest) {

        Post post = postRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Post not found"));

        post.setContent(postRequest.getContent());
        postRepository.save(post);
        return postMapper.toDTO((post));
    }

    @CacheEvict(value = {"posts", "postsList"}, key = "#id", allEntries = true)
    public void deletePost(Long id) {
        if (!postRepository.existsById(id)) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Post not found");

        }
        postRepository.deleteById(id);
    }


}

File: 251003/Pankratiev/rest/publisher/src/main/kotlin/bsuir/dc/publisher/service/PostService.kt
package bsuir.dc.publisher.service

import bsuir.dc.publisher.dto.from.PostRequestTo
import bsuir.dc.publisher.dto.to.PostResponseTo
import bsuir.dc.publisher.mapper.toEntity
import bsuir.dc.publisher.mapper.toResponse
import bsuir.dc.publisher.repository.IssueRepository
import bsuir.dc.publisher.repository.PostRepository
import org.springframework.stereotype.Service

@Service
class PostService(
    private val postRepository: PostRepository,
    private val issueRepository: IssueRepository,
) {
    fun createPost(postRequestTo: PostRequestTo): PostResponseTo {
        val issue = issueRepository.findById(postRequestTo.issueId).orElseThrow { NoSuchElementException() }
        val post = postRequestTo.toEntity(issue)
        val savedPost = postRepository.save(post)
        return savedPost.toResponse()
    }

    fun getPostById(id: Long): PostResponseTo {
        val post = postRepository.findById(id).orElseThrow { NoSuchElementException() }
        return post.toResponse()
    }

    fun getAllPosts(): List<PostResponseTo> =
        postRepository.findAll().map { it.toResponse() }

    fun updatePost(id: Long, postRequestTo: PostRequestTo): PostResponseTo {
        val issue = issueRepository.findById(postRequestTo.issueId).orElseThrow { NoSuchElementException() }
        val updatedPost = postRequestTo.toEntity(issue).apply { this.id = id }
        return postRepository.save(updatedPost).toResponse()
    }

    fun deletePost(id: Long) {
        postRepository.findById(id).orElseThrow { NoSuchElementException() }
        postRepository.deleteById(id)
    }

    fun getPostsByIssueId(issueId: Long): List<PostResponseTo> {
        val issue = issueRepository.findById(issueId).orElseThrow { NoSuchElementException() }
        return issue.posts.map { it.toResponse() }
    }
}